pkg/controllers/sync/controller.go

```diff
@@ -82,242 +88,241 @@ const (
-		controllerConfig.Metrics,
-		deliverutil.NewMetricTags("sync-worker", typeConfig.GetTargetType().Kind),
+		workerCount,
+		metrics,
 	)
 
-	// TODO: do we need both clusterWorker and clusterQueue?
-	s.clusterWorker = worker.NewReconcileWorker(s.reconcileCluster, worker.RateLimiterOptions{}, 1, controllerConfig.Metrics,
-		deliverutil.NewMetricTags("sync-cluster-worker", typeConfig.GetTargetType().Kind))
+	s.clusterCascadingDeletionWorker = worker.NewReconcileWorker[common.QualifiedName](
+		SyncControllerName+"-cluster-cascading-deletion-worker",
+		nil,
+		s.reconcileClusterForCascadingDeletion,
+		worker.RateLimiterOptions{},
+		1,
+		metrics,
+	)
 
-	// Build queue for triggering cluster reconciliations.
-	s.clusterQueue = workqueue.NewNamedDelayingQueue("sync-controller-cluster-queue")
+	// Build queue for triggering reconciliation of all federated resources..
+	s.reconcileAllResourcesQueue = workqueue.NewNamedDelayingQueue(SyncControllerName + "-reconcile-all-resources-queue")
 
-	targetAPIResource := typeConfig.GetTargetType()
+	if err := s.ftcManager.AddFTCUpdateHandler(func(lastObserved, latest *fedcorev1a1.FederatedTypeConfig) {
+		isNewFTC := lastObserved == nil && latest != nil
+		ftcPathDefinitionsChanged := lastObserved != nil && latest != nil && lastObserved.Spec.PathDefinition != latest.Spec.PathDefinition
+		if isNewFTC || ftcPathDefinitionsChanged {
+			s.enqueueForGVK(latest.GetSourceTypeGVK())
+		}
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add FTC update handler: %w", err)
+	}
+
+	if err := s.fedInformerManager.AddEventHandlerGenerator(&informermanager.EventHandlerGenerator{
+		Predicate: informermanager.RegisterOncePredicate,
+		Generator: func(ftc *fedcorev1a1.FederatedTypeConfig) cache.ResourceEventHandler {
+			return eventhandlers.NewTriggerOnAllChanges(func(o pkgruntime.Object) {
+				obj := o.(*unstructured.Unstructured)
+
+				ftc, exists := s.ftcManager.GetResourceFTC(obj.GroupVersionKind())
+				if !exists {
+					return
+				}
+
+				federatedName := common.QualifiedName{
+					Namespace: obj.GetNamespace(),
+					Name:      naming.GenerateFederatedObjectName(obj.GetName(), ftc.GetName()),
+				}
+				s.worker.EnqueueWithDelay(federatedName, s.memberObjectEnqueueDelay)
+			})
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add event handler generator: %w", err)
+	}
 
-	// Federated informer for resources in member clusters
-	s.informer, err = util.NewFederatedInformer(
-		controllerConfig,
-		client,
-		configCopy,
-		&targetAPIResource,
-		func(obj pkgruntime.Object) {
-			qualifiedName := common.NewQualifiedName(obj)
-			s.worker.EnqueueWithDelay(qualifiedName, s.memberObjectEnqueueDelay)
+	if err := s.fedInformerManager.AddClusterEventHandlers(
+		&informermanager.ClusterEventHandler{
+			Predicate: func(oldCluster, newCluster *fedcorev1a1.FederatedCluster) bool {
+				// Enqueue cluster when it's added or marked for deletion to ensure cascading deletion
```

Comment:
Will this cause us to mistakenly enqueue unjoined clusters?

Reply:
Does this also overlap with the predicate below?

> // Trigger cascading deletion when cluster is marked for deletion

Reply:
We will enqueue unjoined clusters, but `reconcileClusterForCascadingDelete` will handle them correctly.